File: swift/Sources/WalletTestKit/Contracts.swift
import CborCose
import WalletAPI

/*
 * Port contract test suites (framework-agnostic).
 *
 * "어댑터 자격 = 계약 테스트 통과": the same checks run against SoftwareSecureArea /
 * InMemoryStorageDriver on Linux CI and against real adapters in the device lab.
 */

public struct ContractViolation: Error, CustomStringConvertible {
    public let description: String

    init(_ description: String) {
        self.description = description
    }
}

private func require(_ condition: Bool, _ message: @autoclosure () -> String) throws {
    if !condition { throw ContractViolation(message()) }
}

public enum SecureAreaContract {

    public static func verify(_ area: any SecureArea) async throws {
        try require(!area.capabilities.algorithms.isEmpty, "capabilities must declare at least one algorithm")

        for algorithm in area.capabilities.algorithms {
            let info = try await area.createKey(spec: KeySpec(secureArea: area.id, algorithm: algorithm))
            try require(info.handle.secureArea == area.id, "\(algorithm): handle must reference this area")
            try require(info.algorithm == algorithm, "\(algorithm): KeyInfo.algorithm mismatch")

            let fetched = try await area.publicKey(key: info.handle)
            try require(
                fetched.x == info.publicKey.x && fetched.y == info.publicKey.y,
                "\(algorithm): publicKey(handle) must equal createKey's public key"
            )

            let data = [UInt8]("port-contract-test".utf8)
            let signature = try await area.sign(key: info.handle, algorithm: algorithm, data: data, hint: nil)
            try require(
                signature.count == 2 * algorithm.curve.coordinateSize,
                "\(algorithm): signature must be raw r||s, got \(signature.count) bytes"
            )
            try require(
                Ecdsa.verify(key: info.publicKey, algorithm: algorithm.coseAlgorithm, data: data, rawSignature: signature),
                "\(algorithm): signature must verify against the key's public key"
            )
            try require(
                !Ecdsa.verify(key: info.publicKey, algorithm: algorithm.coseAlgorithm, data: [UInt8]("other-data".utf8), rawSignature: signature),
                "\(algorithm): signature must not verify for different data"
            )

            if area.capabilities.keyAttestation {
                let challenge = (0..<32).map { UInt8($0) }
                let attestation = try await area.attestation(key: info.handle, challenge: challenge)
                try require(attestation != nil, "\(algorithm): area declares keyAttestation but returned nil")
            }

            try await area.deleteKey(key: info.handle)
            var signAfterDeleteFailed = false
            do {
                _ = try await area.sign(key: info.handle, algorithm: algorithm, data: data, hint: nil)
            } catch {
                signAfterDeleteFailed = true
            }
            try require(signAfterDeleteFailed, "\(algorithm): signing with a deleted key must fail")
        }

        if area.capabilities.keyAgreement {
            let algorithm = area.capabilities.algorithms.first!
            let a = try await area.createKey(spec: KeySpec(secureArea: area.id, algorithm: algorithm))
            let b = try await area.createKey(spec: KeySpec(secureArea: area.id, algorithm: algorithm))
            let s1 = try await area.keyAgreement(key: a.handle, peerPublicKey: b.publicKey, hint: nil)
            let s2 = try await area.keyAgreement(key: b.handle, peerPublicKey: a.publicKey, hint: nil)
            try require(!s1.isEmpty && s1 == s2, "ECDH must be symmetric and non-empty")
            try await area.deleteKey(key: a.handle)
            try await area.deleteKey(key: b.handle)
        }
    }
}

public enum StorageDriverContract {

    public static func verify(_ driver: any StorageDriver) async throws {
        let c = "contract-test"

        try require(try await driver.get(collection: c, key: "missing") == nil, "missing key must read as nil")

        try await driver.put(collection: c, key: "k1", value: [1, 2, 3])
        try require(try await driver.get(collection: c, key: "k1") == [1, 2, 3], "put/get roundtrip")

        try await driver.put(collection: c, key: "k1", value: [9])
        try require(try await driver.get(collection: c, key: "k1") == [9], "overwrite must replace value")

        try await driver.put(collection: c, key: "k2", value: [2])
        try require(Set(try await driver.keys(collection: c)) == ["k1", "k2"], "keys() must list stored keys")

        try await driver.delete(collection: c, key: "k1")
        try require(try await driver.get(collection: c, key: "k1") == nil, "delete must remove the value")

        try await driver.transaction { tx in
            try await tx.put(collection: c, key: "k3", value: [3])
            guard try await tx.get(collection: c, key: "k2") == [2] else {
                throw ContractViolation("tx must see pre-existing state")
            }
            guard try await tx.get(collection: c, key: "k3") == [3] else {
                throw ContractViolation("tx must see its own writes")
            }
            try await tx.delete(collection: c, key: "k2")
        }
        try require(try await driver.get(collection: c, key: "k3") == [3], "tx writes must persist")
        try require(try await driver.get(collection: c, key: "k2") == nil, "tx deletes must persist")

        try await driver.delete(collection: c, key: "k3")
    }
}

File: kotlin/testkit/src/main/kotlin/com/hopae/eudi/wallet/testkit/Contracts.kt
package com.hopae.eudi.wallet.testkit

import com.hopae.eudi.wallet.cbor.cose.Ecdsa
import com.hopae.eudi.wallet.spi.KeySpec
import com.hopae.eudi.wallet.spi.SecureArea
import com.hopae.eudi.wallet.spi.StorageDriver
import com.hopae.eudi.wallet.spi.coseAlgorithm
import com.hopae.eudi.wallet.spi.curve

/*
 * Port contract test suites (framework-agnostic).
 *
 * "어댑터 자격 = 계약 테스트 통과": the same checks run against SoftwareSecureArea /
 * InMemoryStorageDriver on Linux CI and against real adapters in the device lab.
 * Failures throw IllegalStateException with a descriptive message.
 */

object SecureAreaContract {

    suspend fun verify(area: SecureArea) {
        check(area.capabilities.algorithms.isNotEmpty()) { "capabilities must declare at least one algorithm" }

        for (algorithm in area.capabilities.algorithms) {
            val info = area.createKey(KeySpec(secureArea = area.id, algorithm = algorithm))
            check(info.handle.secureArea == area.id) { "$algorithm: handle must reference this area" }
            check(info.algorithm == algorithm) { "$algorithm: KeyInfo.algorithm mismatch" }

            val fetched = area.publicKey(info.handle)
            check(
                fetched.x.contentEquals(info.publicKey.x) && fetched.y.contentEquals(info.publicKey.y)
            ) { "$algorithm: publicKey(handle) must equal createKey's public key" }

            val data = "port-contract-test".encodeToByteArray()
            val signature = area.sign(info.handle, algorithm, data, null)
            check(signature.size == 2 * algorithm.curve.coordinateSize) {
                "$algorithm: signature must be raw r||s (${2 * algorithm.curve.coordinateSize} bytes), got ${signature.size}"
            }
            check(Ecdsa.verify(info.publicKey, algorithm.coseAlgorithm, data, signature)) {
                "$algorithm: signature must verify against the key's public key"
            }
            check(!Ecdsa.verify(info.publicKey, algorithm.coseAlgorithm, "other-data".encodeToByteArray(), signature)) {
                "$algorithm: signature must not verify for different data"
            }

            if (area.capabilities.keyAttestation) {
                checkNotNull(area.attestation(info.handle, ByteArray(32) { it.toByte() })) {
                    "$algorithm: area declares keyAttestation but returned null"
                }
            }

            area.deleteKey(info.handle)
            check(runCatching { area.sign(info.handle, algorithm, data, null) }.isFailure) {
                "$algorithm: signing with a deleted key must fail"
            }
        }

        if (area.capabilities.keyAgreement) {
            val algorithm = area.capabilities.algorithms.first()
            val a = area.createKey(KeySpec(secureArea = area.id, algorithm = algorithm))
            val b = area.createKey(KeySpec(secureArea = area.id, algorithm = algorithm))
            val s1 = area.keyAgreement(a.handle, b.publicKey, null)
            val s2 = area.keyAgreement(b.handle, a.publicKey, null)
            check(s1.isNotEmpty() && s1.contentEquals(s2)) { "ECDH must be symmetric and non-empty" }
            area.deleteKey(a.handle)
            area.deleteKey(b.handle)
        }
    }
}

object StorageDriverContract {

    suspend fun verify(driver: StorageDriver) {
        val c = "contract-test"

        check(driver.get(c, "missing") == null) { "missing key must read as null" }

        driver.put(c, "k1", byteArrayOf(1, 2, 3))
        check(driver.get(c, "k1")?.contentEquals(byteArrayOf(1, 2, 3)) == true) { "put/get roundtrip" }

        driver.get(c, "k1")!![0] = 99
        check(driver.get(c, "k1")?.get(0) == 1.toByte()) { "returned values must not alias stored state" }

        driver.put(c, "k1", byteArrayOf(9))
        check(driver.get(c, "k1")?.contentEquals(byteArrayOf(9)) == true) { "overwrite must replace value" }

        driver.put(c, "k2", byteArrayOf(2))
        check(driver.keys(c).toSet() == setOf("k1", "k2")) { "keys() must list stored keys" }

        driver.delete(c, "k1")
        check(driver.get(c, "k1") == null) { "delete must remove the value" }

        driver.transaction {
            put(c, "k3", byteArrayOf(3))
            check(get(c, "k2")?.contentEquals(byteArrayOf(2)) == true) { "tx must see pre-existing state" }
            check(get(c, "k3")?.contentEquals(byteArrayOf(3)) == true) { "tx must see its own writes" }
            delete(c, "k2")
        }
        check(driver.get(c, "k3")?.contentEquals(byteArrayOf(3)) == true) { "tx writes must persist" }
        check(driver.get(c, "k2") == null) { "tx deletes must persist" }

        driver.delete(c, "k3")
    }
}
